### DISCUSSION TITLE
NativeProcessStartupInfo filepaths with spaces arguments

### DISCUSSION THREAD

**Discussion by MalacTheLittle** [airsdk/Adobe-Runtime-Support#1295] [Q&A]
I'm trying to use external 7zip app to extract some big files from zip archive on Windows. Both, executable and archive files has spaces in filepath:
E:\ELEMENT TEST\appData\7za.exe
E:\ZIP FILES\book package.zip

I set executable file "C:Windows\System32\cmd.exe" and pass arguments in order:
["/c", "E:\ELEMENT TEST\appData\7za.exe", "e", "E:\ZIP FILES\book package.zip"],
I'm getting `standardError` "'E:\ELEMENT' is not recognized as an internal or external command, operable program or batch file." which means it breaks at first space found in arguments.

Now, when I pass only:
["/c", "E:\ELEMENT TEST\appData\7za.exe"] ,
it works fine and I get `standardOutput` with .exe "help" instructions. 

If I change path to 7za.exe to be without spaces, everything goes without a problem.
So, if I push 2 or more paths with spaces to arguments, I'm getting error!

I also tried to set working directory to either .exe or .zip path, still getting same results.

As far as I was able to find, the problem dates back to 2011 ([here](https://stackoverflow.com/questions/6228515/adobe-air-nativeprocess-fails-with-spaces-in-arguments)).

Is there any solution for this or I should report an issue?

**Comment by ajwfrost** [comment]
I'm hoping that you're just putting paths there as you would see them in Windows, rather than how you would write them in an AS3 string? As the `\` character is the escape character .. so `E:\ZIP FILES\book package.zip` would need to be written as `"E:\\ZIP FILES\\book package.zip"`

Have you tried then enclosing the argument in quotes, i.e. `"\"E:\\ZIP FILES\\book package.zip\""`

Interestingly there's a function in the NativeProcess ActionScript definition that will try to do this sort of thing: if any space is found then it encloses the whole argument in quote marks. But it also delimits other things that would be needed when we are passing a string to an operating system so if you had a string that itself contained a quote, it would turn that into an escaped quote mark. This function is actually then never invoked, so my guess would be that it caused too many side effects for people who were already working around the issue by putting their own quotes on!

So, for arguments, if they do have spaces then please wrap them in quotes (and on a mac I guess you would need to do the same but with forward-slashes e.g. `"\"/Users/me/my file.zip\"`)

thanks

**Reply by MalacTheLittle** [reply]
You're right, I tried adding quotes and escaping quotes and escaping backslash separators with no success. I think the problem lays in automatic escaping of arguments that are pushed to `NativeProcessStartupInfo` as it's stated at Adobe API ([here](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/desktop/NativeProcessStartupInfo.html#arguments)): "AIR automatically escapes any characters in the string that need to be escaped (such as space characters)." Therefore, additional escaping of spaces and backslashes doesn't help :/

**Reply by MalacTheLittle** [reply]
Here are examples what I've tried so far:
Without adding any quotes `['/c', zipExe.nativePath, 'e', zipFile.nativePath]` I get `standardError`:
'E:\ELEMENT' is not recognized as an internal or external command, operable program or batch file.

When I add quotes `['/c', '"' + zipExe.nativePath + '"', 'e', '"' + zipFile.nativePath + '"']` I get `standarError`:
'\\"E:\ELEMENT TEST\7za.exe\\"" e "\\"E:\ZIP FILES\book package.zip\\"' is not recognized as an internal or external command, operable program or batch file.

If I add escaped quotes `['/c', '\"' + zipExe.nativePath + '\"', 'e', '\"' + zipFile.nativePath + '\"']` I get `standarError`:
'\\"E:\ELEMENT TEST\7za.exe\"" e "\\"E:\ZIP FILES\book package.zip\"' is not recognized as an internal or external command, operable program or batch file.

If I add escaped quotes and escape escapes `['/c', '\\"' + zipExe.nativePath + '\\"', 'e', '\\"' + zipFile.nativePath + '\\"']` I get `standarError`:
'\\\\\\"E:\ELEMENT TEST\7za.exe\\\\\\"" e "\\\\\\"E:\ZIP FILES\book package.zip\\\\\\"' is not recognized as an internal or external command, operable program or batch file.

Actually, there is no way to pass quotes since it gets automatically escaped by `NativeProcessStartupInfo`.

**Reply by marchbold** [reply]
I think when you are using cmd you need to remember there are double set of arguments i.e you are passing a set of arguments as a command to the `cmd.exe` so it's a little more complex than just a straight set of arguments. Probably why you are finding using the exe directly works better. 

It's worth checking the documentation of `cmd.exe` as there are certain parameters that need two quotes, `""demo.exe""` 

Here is one of the examples with spaces in program path and params:

```
cmd /c ""c:\path with spaces\demo.exe" "Parameter 1 with space" "Parameter2 with space""
```

You'll notice it's one argument, rather than multiple. So could be worth trying it like that?

**Reply by MalacTheLittle** [reply]
Good point, I didn't think of that, still same result :/

First I tryied as a single argument:
`['/c', zipExe.nativePath + ' e ' + zipFile.nativePath]`
'E:\ELEMENT' is not recognized as an internal or external command, operable program or batch file.

Then as a single argument with quotes:
`['/c', '"' + zipExe.nativePath + ' e ' + zipFile.nativePath + '"']`
'\"E:\ELEMENT TEST\7za.exe e E:\ZIP FILES\book package.zip\"' is not recognized as an internal or external command, operable program or batch file.

There are two problems with `cmd.exe`: (a) filenames/foldernames with spaces has to be wrapped with quotes, (b) it doesn't accept single quotes. And I can't pass unescaped quotes :/

**Reply by marchbold** [reply]
Did you try like this:

```
['/c', '""' + zipExe.nativePath + '" e "' + zipFile.nativePath + '""']
```

**Reply by MalacTheLittle** [reply]
Now I get:
The filename, directory name, or volume label syntax is incorrect.

And sorry, I've overlooked that I have to escape backslashes here too :D
So in preview post second result example looks like this:
`['/c', '"' + zipExe.nativePath + ' e ' + zipFile.nativePath + '"']`
`'\"E:\ELEMENT TEST\7za.exe e E:\ZIP FILES\book package.zip\"' is not recognized as an internal or external command, operable program or batch file.`
Actually, any quote in any form is being escaped :/

**Reply by marchbold** [reply]
Did you try it with the two quotes?

**Reply by MalacTheLittle** [reply]
Yes, then I get: The filename, directory name, or volume label syntax is incorrect.
Obiously, the argument that cmd gets is `\"\"E:\ELEMENT TEST\7za.exe\" e \"E:\ZIP FILES\book package.zip\"\"` since quotes are escaped.

**Reply by marchbold** [reply]
Yeah definitely looks like an escaping issue...

**Comment by MalacTheLittle** [comment]
OK, I found solution for my case:
I set 7za.exe as executable instead of cmd.exe and pass arguments like this: 
`['e', zipFile.nativePath, 'Some file in archive with spaces.xml', '-o' + zipFile.parent.nativePath, '-y']`
and it works like a charm for many arguments with spaces :)

And FYI, zipFile.nativePath is E:\ZIP FILES\book package.zip
I think that 7za.exe has better arguments processing than cmd.exe. 

Maybe possible solution for quotes would be to add property for disabling escaping of arguments, something like `NativeProcessStartupInfo.disableArgumentsEscape = true;` That way we can escape only parts that has to be escaped :)
